src/s3/multipartcopy.rs

```diff
@@ -0,0 +1,311 @@
+use aws_sdk_s3::error::SdkError;
+use aws_sdk_s3::operation::complete_multipart_upload::CompleteMultipartUploadError;
+use aws_sdk_s3::operation::create_multipart_upload::{
+    CreateMultipartUploadError, CreateMultipartUploadOutput,
+};
+use aws_sdk_s3::operation::head_object::HeadObjectError;
+use aws_sdk_s3::types::{CompletedMultipartUpload, CompletedPart};
+use aws_sdk_s3::Client;
+use bytesize::MIB;
+use conv::{prelude::*, FloatError, RangeError};
+use futures::stream::{self, StreamExt, TryStreamExt};
+use std::error::Error as StdError;
+use std::num::TryFromIntError;
+use std::sync::Arc;
+use thiserror::Error;
+use tracing::instrument;
+use typed_builder::TypedBuilder;
+
+use super::S3Object;
+
+/// Retrieves the size of the source object in the specified S3 bucket.
+///
+/// # Arguments
+///
+/// * `client` - A reference to the S3 client.
+/// * `bucket` - The name of the S3 bucket.
+/// * `key` - The key of the source object.
+///
+/// # Returns
+///
+/// The size of the source object in bytes.
+#[instrument(skip(client))]
+pub async fn get_source_size(
+    client: &Client,
+    bucket: &str,
+    key: &str,
+) -> Result<i64, aws_sdk_s3::error::SdkError<HeadObjectError>> {
+    let head_object = client.head_object().bucket(bucket).key(key).send().await?;
+
+    Ok(head_object.content_length().unwrap_or(0))
+}
+
+const DEFAULT_COPY_PART_SIZE: i64 = 50 * MIB as i64;
+
+/// Custom error types for S3 multipart copy operations.
+#[derive(Debug, Error)]
+pub enum S3MultipartCopierError {
+    #[error("Missing multipart upload id")]
+    MissingUploadId,
+    #[error("Missing copy part result")]
+    MissingCopyPartResult,
+    #[error(transparent)]
+    S3Error(Box<dyn StdError + Send + Sync>),
+    #[error(transparent)]
+    RangeError(#[from] RangeError<i64>),
+    #[error(transparent)]
+    FloatError(#[from] FloatError<f64>),
+    #[error(transparent)]
+    TryFromIntError(#[from] TryFromIntError),
+}
+
+impl<E: StdError + Send + Sync + 'static> From<SdkError<E>> for S3MultipartCopierError {
+    fn from(value: SdkError<E>) -> Self {
+        Self::S3Error(Box::new(value))
+    }
+}
+
+/// A struct representing the parameters required for copying a part of an S3 object.
+#[derive(Debug, TypedBuilder)]
+struct CopyUploadPart<'a> {
+    src: &'a S3Object,
+    dst: &'a S3Object,
+    upload_id: &'a str,
+    part_number: i32,
+    byte_range: (i64, i64),
+}
+
+/// A struct to handle S3 multipart copy operations.
```

Comment:
**suggestion:** Could we add some more docs on how to use this feature? It looks like we get some docs for free from `TypedBuilder`, but they're not great. 

<img width="1043" alt="Screenshot 2024-06-03 at 11 07 19 AM" src="https://github.com/harrison-ai/cobalt-aws/assets/616382/76f324e2-8097-49cf-b179-a123f4cd92a7">

In particular, there's no documentation of the individual fields, their types, or how they should be used in the builder. I guess we could work around this by making the fields `pub` and adding docstrings, or else perhaps an example, like below (but formatted correctly 😅).

In particular, it would be helpful to have some guidance on the `part_size` and `max_concurrent_uploads` fields, and when/why the user might want to set them.

```suggestion
/// A struct to handle S3 multipart copy operations.
///
///        let dest = S3Object::new(test_bucket, dst_key);
///        let copyier = S3MultipartCopier::builder()
///            .client(client.clone())
///            .source(src)
///            .destination(dest.clone())
///            .part_size((5 * MIB).try_into()?)
///            .max_concurrent_uploads(2)
///            .build();
///        copyier.send().await?;
```

You can run the following command to generate the docs locally
```sh
RUSTDOCFLAGS="--cfg docsrs" cargo +nightly doc --open --all-features
```

Reply:
> However, it's important to note the potential increase in costs due to the higher number of API requests made by multipart copy operations.

Would be useful to leave a comment for the user about this in this documentation block 👍